### ISSUE TITLE
How to cache NEventStore using annotation?

### ISSUE THREAD

**Issue by leenux** [NEventStore/NEventStore#427]
Hi,
I'm a new for NEventStore,I dont know how to cache NEventStore using annotation.

Example for method:
[NEventStore.cache=true,...etc.]
getPerson(...);

Or for object or member:
[NEventStore.cache=true,...etc.]
class Person{};

Thank you.

**Comment by bartelink** [comment]
`NEventStore` does not have any behaviors of this nature.

The closest thing I can think of is that `EventStoreRepository`has some caching behaviors, but in general with Event Sourcing, you're far better off getting something stood up which works correctly based on loading the data from fresh and then refactor to cache as necessary from there.

If you read [between the lines] the NEventStore docs, one of the key considerations in any such caching scheme are to the mechanism for managing concurrent access is based on an optimistic scheme. Any caching mechanism needs to [aside from providing sufficient perf benefit] also interact correctly with the overall scheme. In other words, if it turns out that the cached value was too stale you'd need to trigger reloading and/or use the Conflict Resolution mechanisms to deal with it appropriately.

The bottom line is that I'd stay well away from such optimisations until you've determined that NEventStore is sufficiently capable and 'alive' for your purposes - there are lots of things to udnerstand before you get to considering caching.

**Comment by AGiorgetti** [comment]
NEventStore at the moment does not have explicit support for caching, if you look at how the repository classes are implemented here: 

https://github.com/NEventStore/NEventStore.Domain/blob/master/src/NEventStore.Domain/Persistence/EventStore/EventStoreRepository.cs

You'll notice two dictionaries, one for streams and one for snapshots, if you want a rudimentary caching mechanic just keep those repositories as singleton, but I'll not recommend that due to potential thread safety and concurrency issues.

We'll review a caching mechanism in the future, suggestions are always welcome.
